nexus/src/app/sagas/region_snapshot_replacement_start.rs

```diff
@@ -0,0 +1,1161 @@
+        log,
+        "replacing {} with {} in volume {}",
+        old_snapshot_address,
+        new_region_address,
+        old_volume_id,
+    );
+
+    // `volume_replace_snapshot` will swap the old snapshot for the new region.
+    // No repair or reconcilation needs to occur after this.
+    osagactx
+        .datastore()
+        .volume_replace_snapshot(
+            VolumeWithTarget(old_volume_id),
+            ExistingTarget(old_snapshot_address),
+            ReplacementTarget(new_region_address),
+            VolumeToDelete(new_volume_id),
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(())
+}
+
+async fn rsrss_replace_snapshot_in_volume_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    // Undo the forward action's volume_replace_snapshot call by swapping the
+    // target_addr fields in the parameters.
+
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+
+    let new_volume_id = sagactx.lookup::<Uuid>("new_volume_id")?;
+
+    let region_snapshot = osagactx
+        .datastore()
+        .region_snapshot_get(
+            params.request.old_dataset_id,
+            params.request.old_region_id,
+            params.request.old_snapshot_id,
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let Some(region_snapshot) = region_snapshot else {
+        bail!(
+            "region snapshot {} {} {} deleted!",
+            params.request.old_dataset_id,
+            params.request.old_region_id,
+            params.request.old_snapshot_id,
+        );
+    };
+
+    let old_snapshot_address: SocketAddrV6 =
+        match region_snapshot.snapshot_addr.parse() {
+            Ok(addr) => addr,
+
+            Err(e) => {
+                bail!(
+                    "parsing {} as SocketAddrV6 failed: {e}",
+                    region_snapshot.snapshot_addr,
+                );
+            }
+        };
+
+    let (new_dataset, ensured_region) = sagactx.lookup::<(
+        db::model::Dataset,
+        crucible_agent_client::types::Region,
+    )>(
+        "ensured_dataset_and_region",
+    )?;
+
+    let Some(new_dataset_address) = new_dataset.address() else {
+        bail!("dataset {} does not have an address!", new_dataset.id());
+    };
+
+    let new_region_address = SocketAddrV6::new(
+        *new_dataset_address.ip(),
+        ensured_region.port_number,
+        0,
+        0,
+    );
+
+    // The forward action will have overwritten db_region's volume id, so get
+    // the cached copy.
+    let old_volume_id = sagactx.lookup::<Uuid>("old_snapshot_volume_id")?;
+
+    info!(
+        log,
+        "undo: replacing {} with {} in volume {}",
+        old_snapshot_address,
+        new_region_address,
+        old_volume_id,
+    );
+
+    osagactx
+        .datastore()
+        .volume_replace_snapshot(
+            VolumeWithTarget(old_volume_id),
```

Comment:
AFAICT, the only change between this undo and the prior method (as stated in the comment) is this swap in parameters and the log message above. It may be worth while to create a helper method to do all the lookups and then only perform this last call in the forward/undo functions. That would make it clearer what is different IMO.

Reply:
agreed, done in 94bc3cbb3